astronomy/standard_product_3.cpp

```diff
@@ -0,0 +1,310 @@
+    CHECK_LT(index - 1, line->size()) << location;
+    return (*line)[index - 1];
+  };
+  auto const columns = [&line, &location](
+                            int first, int last) -> std::string_view {
+    CHECK(line.has_value()) << location;
+    CHECK_LT(last - 1, line->size()) << location;
+    return std::string_view(&(*line)[first - 1], last - first + 1);
+  };
+  auto const float_columns = [&columns, &location](int first,
+                                                   int last) -> double {
+    double result;
+    CHECK(absl::SimpleAtod(columns(first, last), &result))
+        << location << " columns " << first << "-" << last;
+    return result;
+  };
+  auto const integer_columns = [&columns, &location](int first,
+                                                     int last) -> int {
+    int result;
+    CHECK(absl::SimpleAtoi(columns(first, last), &result))
+        << location << " columns " << first << "-" << last;
+    return result;
+  };
+
+  int number_of_epochs;
+  int number_of_satellites;
+
+  // Header: # record.
+  read_line();
+  CHECK_EQ(column(1), '#') << location;
+  CHECK_GE(column(2), 'a') << location;
+  CHECK_LE(column(2), 'd') << location;
+  version_ = column(2);
+  CHECK(column(3) == 'P' || column(3) == 'V') << location;
+  has_velocities_ = column(3) == 'V';
+  number_of_epochs = integer_columns(33, 39);
+  if (dialect == Dialect::ILRSB) {
+    --number_of_epochs;
+  }
+
+  // Header: ## record.
+  read_line();
+  CHECK_EQ(columns(1, 2), "##") << location;
+
+  // Header: +␣ records.
+  read_line();
+  CHECK_EQ(columns(1, 2), "+ ") << location;
+  number_of_satellites = integer_columns(4, 6);
+
+  int number_of_satellite_id_records = 0;
+  while (columns(1, 2) == "+ ") {
+    ++number_of_satellite_id_records;
+    for (int c = 10; c <= 58; c += 3) {
+      auto const full_location =
+          absl::StrCat(location, " columns ", c, "-", c + 2);
+      if (orbits_.size() != number_of_satellites) {
+        SatelliteIdentifier id;
+        if (version_ == 'a') {
+          // Satellite IDs are purely numeric (and implicitly GPS) in SP3-a.
+          CHECK_EQ(column(c), ' ') << full_location;
+          id.group = GPS;
+        } else {
+          id.group = SatelliteGroup{column(c)};
+          switch (id.group) {
+            case GPS:
+            case ГЛОНАСС:
+              // GPS and ГЛОНАСС (G and R) satellite IDs are supported since
+              // SP3-b.
+              break;
+            case General:
+            case Galileo:
+            case 北斗:
+            case 準天頂衛星:
+            case IRNSS:
+              CHECK_GE(version_, 'c') << full_location;
+              break;
+            default:
+              LOG(FATAL) << "invalid satellite identifier " << id << ": "
+                         << full_location;
+          }
+        }
+        id.index = integer_columns(c + 1, c + 2);
+        CHECK_GT(id.index, 0) << full_location;
+        CHECK(orbits_.emplace(std::piecewise_construct,
```

Comment:
Use `std::tie` and `std::ignore`.

Reply:
Let’s try https://en.cppreference.com/w/cpp/language/structured_binding and see what breaks.